### PR TITLE
Add External Validation Rules for Analysis Properties Based on Dynamic Schema

### PR DESCRIPTION
**Pull request by Azher2Ali**
This PR introduces external validation capabilities for analysis properties, allowing for dynamic schemas to specify validation rules that are stored in the database. Each rule specifies a `jsonPath` to identify the property for validation and a URL to an external service for validation logic. During analysis submission, the system will validate specific property values by querying the external service, confirming that the values (e.g., donor IDs) meet the criteria defined by the external registry or ontology.

- Adds support for configurable external validation rules per analysis type.

- Enables database storage of validation rules tied to specific analysis types.

- Implements web request calls to validate property values according to the rules during submission.

This update also introduces external validation checks for analyses during creation and update in SONG. These checks are applied after JSON schema validation and file type validation. For each configured external validation rule on an analysis type, the system will:

**Check Analysis Properties:** Determine if the property at the specified path has a value. If no value exists, the validation is considered passed.

**Validate Using External Source:** If a value is present, SONG constructs a GET request to the external validation URL. The URL includes query parameters:

- studyId: the study identifier the analysis is associated with.

- value: the value of the property being validated.

**Process the External Validation Response:**

- If the value is valid, the external service returns an HTTP 200 status with no response body.

- If invalid, the external service returns an HTTP 400 status and optionally includes a JSON error message explaining why the value is invalid. This message is provided to the submitter for clarity.

- Other HTTP errors indicate a validation failure, and a message is returned to the submitter stating that external validation was unsuccessful, preventing the analysis submission/update.

This update sets the groundwork for calling external validation services, but the actual call to the external service is pending final implementation .We added this feature with an exception that the actual call to the external service is skipped as the service is not up and running and if we start invoking it, this will fail all the validations  #872 #875

### CODE REVIEW

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java [hunk 1]
```diff
@@ -265,11 +267,33 @@ private AnalysisType commitAnalysisType(
...
+
+      if (!analysisSchemaList.isEmpty()) {
+        Optional<AnalysisSchema> latestSchemaOptional =
+            analysisSchemaList.stream()
+                .filter(schema -> schema.getVersion() != null)
+                .max(Comparator.comparingInt(AnalysisSchema::getVersion));
+
+        if (latestSchemaOptional.isPresent()) {
+          AnalysisTypeOptions optionsFromDb = latestSchemaOptional.get().getOptions();
+          fileTypes = optionsFromDb.getFileTypes();
```
Comment (leoraba) [inline song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java:274]: Looks like this check is retrieving the previous `fileTypes` and not used except for logging.

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java [hunk 2]
```diff
@@ -319,7 +343,10 @@ public static AnalysisTypeId resolveAnalysisTypeId(@NonNull AnalysisType analysi
   private AnalysisType convertToAnalysisType(
       AnalysisSchema analysisSchema, boolean hideSchema, boolean unrenderedOnly) {
     AnalysisTypeOptions options = new AnalysisTypeOptions();
-    options.setFileTypes(analysisSchema.getFileTypes());
+    if (analysisSchema.getOptions() != null) {
+      options.setFileTypes(analysisSchema.getOptions().getFileTypes());
+      options.setExternalValidation(analysisSchema.getOptions().getExternalValidation());
+    }
```
Comment (leoraba) [inline song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java]: This setters using getters seems a bit redundant. Guess the variable `options` is not required as we can use `analysisSchema.getOptions()` directly bellow in the code when using the Builder.

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 3]
```diff
@@ -62,23 +68,26 @@ public class ValidationService {
...
     this.analysisTypeService = analysisTypeService;
     this.uploadRepository = uploadRepository;
     this.enforceLatest = enforceLatest;
     this.analysisTypeIdSchema = analysisTypeIdSchemaSupplier.get();
+    this.restTemplate = restTemplate;
   }
 
   @SneakyThrows
-  public Optional<String> validate(@NonNull JsonNode payload) {
+  public Optional<String> validate(@NonNull JsonNode payload, @NonNull String studyId) {
```
Comment (leoraba) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: `studyId` variable is not used in this method.

File: song-server/src/test/java/bio/overture/song/server/service/MockedSubmitTest.java [hunk 4]
```diff
@@ -101,7 +101,7 @@ public void submit_StudyDne_StudyIdDoesNotExist() {
     // Verify
     assertSongError(
         () -> submitService.submit("anyStudy", "anyAnalysisId", false), STUDY_ID_DOES_NOT_EXIST);
-    verify(validationService, never()).validate(isA(JsonNode.class));
+    verify(validationService, never()).validate(isA(JsonNode.class),"anyStudy");
```
Comment (leoraba) [inline song-server/src/test/java/bio/overture/song/server/service/MockedSubmitTest.java]: what's the reason we're passing an study to the method `validate`?

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 5]
```diff
@@ -109,6 +118,43 @@ public Optional<String> validate(@NonNull JsonNode payload) {
...
+      String value = JsonPath.read(payload, "$." + jsonPath);
+
+      if(Objects.isNull(value) || value.isEmpty())
+        return;
+
+      // call the external url using restTemplate
+    }
+  }
+
+  public boolean invokeExternalUrl(String studyId, String url, String value) {
```
Comment (leoraba) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: Don't see where are we calling this method.

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 6]
```diff
@@ -62,23 +68,26 @@ public class ValidationService {
   private final UploadRepository uploadRepository;
   private final boolean enforceLatest;
   private final Schema analysisTypeIdSchema;
+  private final RestTemplate restTemplate;
```
Comment (leoraba) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: is there a way to use rather a `RetryTemplate`? just in case we require retry mechanism

**Review by leoraba — COMMENTED** [review]
So far so good! Just left some comments to consider 📝

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java (hunk 5 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: We are not calling this method as of now, because the external API is not up and running due to which all the validations will fail.

File: song-server/src/test/java/bio/overture/song/server/service/MockedSubmitTest.java (hunk 4 above)
Comment (Azher2Ali) [reply song-server/src/test/java/bio/overture/song/server/service/MockedSubmitTest.java]: As we need to populate `studyId` value for external validation, hence we have added this as a parameter. If required "`anyStudy`" to "`anyStudyId`"

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java (hunk 3 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: It's not used now, but this will be used when we invoke external URL method

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java (hunk 2 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java]: You are right, will change this part of code.

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java (hunk 1 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java:274]: We will be setting the `fileTypes` fetched from previous versions to `options `object

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 7]
```diff
@@ -109,6 +118,43 @@ public Optional<String> validate(@NonNull JsonNode payload) {
...
+
+      // create URL
+      URIBuilder uriBuilder = new URIBuilder(url);
+      uriBuilder.addParameter("studyId", studyId);
+      uriBuilder.addParameter("value",value);
+      // Make the HTTP GET request
+      ResponseEntity<Void> response = restTemplate.getForEntity(uriBuilder.build().toURL().toString(), Void.class);
+
+      // Return true if the response status is 200 OK
+      return response.getStatusCode().is2xxSuccessful();
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: 👍

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java (hunk 6 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: In order to apply Retry mechanism on the `RestTemplate` we should consider Timeout on UI as retry will take considerable amount of time. As a result, if the `RestTemplate` fails. means that the whole operations failed.

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java [hunk 8]
```diff
@@ -265,11 +267,36 @@ private AnalysisType commitAnalysisType(
     if (options != null && CollectionUtils.isNotEmpty(options.getFileTypes())) {
       fileTypes = options.getFileTypes();
     }
+
+    // checking if file types is empty
+    // if the version is new version of the schema , we are checking the previous version allowed
+    // file types
+    // if it is new then it is empty
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java]: This comment is confusing, I don't understand what it is trying to say. Can you review and rephrase it to describe the purpose of the following section?

File: song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java [hunk 9]
```diff
@@ -265,11 +267,36 @@ private AnalysisType commitAnalysisType(
...
+            analysisSchemaList.stream()
+                .filter(schema -> schema.getVersion() != null)
+                .max(Comparator.comparingInt(AnalysisSchema::getVersion));
+
+        if (latestSchemaOptional.isPresent()) {
+          AnalysisTypeOptions optionsFromDb = latestSchemaOptional.get().getOptions();
+          fileTypes = optionsFromDb.getFileTypes();
+        }
+      }
+    }
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/AnalysisTypeService.java:277]: Why does this PR update how the `fileType` options are being set? Wasn't this functionality already in place?

Do we perform the same logic for `externalValidation` rules? (When an analysis is being updated, if no external validation rules are provided we copy the externalValidaiton rules from the latest schema version)

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 10]
```diff
@@ -100,6 +119,7 @@ public Optional<String> validate(@NonNull JsonNode payload) {
         validateFileType(fileTypes, payload);
       }
 
+      // log.info("SCHEMA :- " + analysisType.getSchema());
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: Remove commented code.

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 11]
```diff
@@ -109,6 +129,65 @@ public Optional<String> validate(@NonNull JsonNode payload) {
...
+      validated = invokeExternalUrl(studyId, externalUrl, value);
+    }
+    return validated;
+  }
+
+  public boolean invokeExternalUrl(String studyId, String url, String value) {
+    try {
+
+      Pattern pattern = Pattern.compile(LYRIC_URL_REGEX);
+      Matcher matcher = pattern.matcher(url);
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: Variable names should describe their content, not just `pattern` and `matcher` but `lyricUrlPattern` and `lyricUrlMatcher` or similar.

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java [hunk 12]
```diff
@@ -109,6 +129,65 @@ public Optional<String> validate(@NonNull JsonNode payload) {
...
+      // call the external url using restTemplate
+      validated = invokeExternalUrl(studyId, externalUrl, value);
+    }
+    return validated;
+  }
+
+  public boolean invokeExternalUrl(String studyId, String url, String value) {
+    try {
+
+      Pattern pattern = Pattern.compile(LYRIC_URL_REGEX);
```
Comment (joneubank) [inline song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: What about external validation URLs when they are validating against a service that isn't Lyric? We need to define a schema for writing a URL that isn't specific to Lyric. This is about defining string template rules with variables we will replace with the studyId and entity name. 

Perhaps `${studyId}`, `${entityName}`, and `${entityId}`. So an external URL value could be `https://mydomain.example.org/whatever/path/to/validate/${studyId}/${entityName}/${entityId}`.

File: song-server/src/main/resources/db/migration/V1_21__remove_file_types_and_add_options.sql [hunk 13]
```diff
@@ -0,0 +1,8 @@
+
+-- Remove the file_types column
+ALTER TABLE public.analysis_schema
+DROP COLUMN file_types;
```
Comment (joneubank) [inline song-server/src/main/resources/db/migration/V1_21__remove_file_types_and_add_options.sql]: Have we published this version with file types yet? We would need to migrate all the file_types that have already been set into the new analysis schema format.

Did you consider storign analysis_schema in a separate column instead of storing options as a JSON?

**Review by joneubank — CHANGES_REQUESTED** [review]

File: song-server/src/main/java/bio/overture/song/server/service/ValidationService.java (hunk 12 above)
Comment (Azher2Ali) [reply song-server/src/main/java/bio/overture/song/server/service/ValidationService.java]: The code will take any URL apart from lyric, but for the path and path params should be well defined in order to have a common logic

File: song-server/src/main/resources/db/migration/V1_21__remove_file_types_and_add_options.sql (hunk 13 above)
Comment (joneubank) [reply song-server/src/main/resources/db/migration/V1_21__remove_file_types_and_add_options.sql]: I think that in the long term it would be best to store all options as their own dedicated columns in postgres. For the scope of this PR, we'll just leave it as a column for options since this is working today.
